presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryPagesStore.java

```diff
@@ -0,0 +1,130 @@
+import static com.facebook.presto.plugin.memory.MemoryErrorCode.MEMORY_LIMIT_EXCEEDED;
+import static com.facebook.presto.plugin.memory.MemoryErrorCode.MISSING_DATA;
+import static java.lang.String.format;
+
+@ThreadSafe
+public class MemoryPagesStore
+{
+    private final long maxBytes;
+
+    @GuardedBy("this")
+    private long currentBytes = 0;
+
+    public MemoryPagesStore(DataSize maxDataSizePerNode)
+    {
+        this.maxBytes = maxDataSizePerNode.toBytes();
+    }
+
+    private final Map<Long, List<Page>> pages = new HashMap<>();
+
+    public synchronized void initialize(long tableId)
+    {
+        if (!pages.containsKey(tableId)) {
+            pages.put(tableId, new ArrayList<>());
+        }
+    }
+
+    public synchronized void add(Long tableId, Page page)
+    {
+        if (!contains(tableId)) {
+            throw new PrestoException(MISSING_DATA, "Failed to find table on a worker.");
+        }
+
+        long newSize = currentBytes + page.getRetainedSizeInBytes();
+        if (maxBytes < newSize) {
+            throw new PrestoException(MEMORY_LIMIT_EXCEEDED, format("Memory limit [%d] for memory connector exceeded", maxBytes));
+        }
+        currentBytes = newSize;
+
+        List<Page> tablePages = pages.get(tableId);
+        tablePages.add(page);
+    }
+
+    public synchronized List<Page> getPages(Long tableId, int partNumber, int totalParts, List<Integer> columnIndexes)
+    {
+        if (!contains(tableId)) {
+            throw new PrestoException(MISSING_DATA, "Failed to find table on a worker.");
+        }
+
+        List<Page> tablePages = pages.get(tableId);
+        ImmutableList.Builder<Page> partitionedPages = ImmutableList.builder();
+
+        for (int i = partNumber; i < tablePages.size(); i += totalParts) {
+            partitionedPages.add(getColumns(tablePages.get(i), columnIndexes));
+        }
+
+        return partitionedPages.build();
+    }
+
+    public synchronized boolean contains(Long tableId)
+    {
+        return pages.containsKey(tableId);
+    }
+
+    public synchronized void cleanUp(Set<Long> activeTableIds)
+    {
+        // We have to remember that there might be some race conditions when there are two tables created at once.
+        // That can lead to a situation when MemoryPagesStore already knows about a newer second table on some worker
+        // but cleanUp is triggered by insert from older first table, which MemoryTableHandle was created before
+        // second table creation. Thus activeTableIds can have missing latest ids and we can only clean up tables
+        // that:
+        // - have smaller value then max(activeTableIds).
+        // - are missing from activeTableIds set
+
+        if (activeTableIds.isEmpty()) {
+            // if activeTableIds is empty, we can not determine latestTableId...
+            return;
+        }
+        long latestTableId  = Collections.max(activeTableIds);
+        for (Long tableId : ImmutableList.copyOf(pages.keySet())) {
```

Comment:
As you wish, however standard practice is utterly ugly and unreadable because of all that boilerplate:

```
        for (Iterator<Map.Entry<Long, List<Page>>> tablePages = pages.entrySet().iterator(); tablePages.hasNext();) {
            Map.Entry<Long, List<Page>> tablePagesEntry = tablePages.next();
            Long tableId = tablePagesEntry.getKey();
            if (tableId < latestTableId && !activeTableIds.contains(tableId)) {
                for (Page removedPage : tablePagesEntry.getValue()) {
                    currentBytes -= removedPage.getRetainedSizeInBytes();
                }
                tablePages.remove();
            }
        }
```